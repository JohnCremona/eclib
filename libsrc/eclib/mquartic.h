// mquartic.h:   Declaration of class quartic and related functions
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
#ifndef _ECLIB_MQUARTIC_H
#define _ECLIB_MQUARTIC_H

#include "points.h"

inline ZZ II(const ZZ& a, const ZZ& b, const ZZ& c, 
	  const ZZ& d, const ZZ& e)
{
  return 12*a*e - 3*b*d + c*c;
}

inline ZZ JJ(const ZZ& a, const ZZ& b, const ZZ& c, 
	  const ZZ& d, const ZZ& e)
{
  return (72*a*e + 9*b*d - 2*sqr(c)) * c - 27*(a*sqr(d) + sqr(b)*e);
}

inline ZZ H_invariant(const ZZ& a, const ZZ& b, const ZZ& c)
{
  return 8*a*c - 3*sqr(b);
}

inline ZZ R_invariant(const ZZ& a, const ZZ& b, const ZZ& c, 
		 const ZZ& d)
{
  return pow(b,3) + 8*d*sqr(a) - 4*a*b*c;
}


class quartic {
public:
     // constructors
        quartic();
        quartic(const ZZ& qa, const ZZ& qb, const ZZ& qc, 
		const ZZ& qd, const ZZ& qe, 
                const vector<bigcomplex>& qr,	int qt,
		const ZZ& qi,const ZZ& qj,const ZZ& qdisc);
        quartic(const ZZ& qa, const ZZ& qb, const ZZ& qc, 
		const ZZ& qd, const ZZ& qe);
  // latter calls set_roots_and_type()
        quartic(const quartic& q);
  // member functions & operators
        void set_roots_and_type();
	void assign(const ZZ& qa, const ZZ& qb, const ZZ& qc, 
		    const ZZ& qd, const ZZ& qe, 
		    const vector<bigcomplex>& qr,    int qt,
		    const ZZ& qi,const ZZ& qj,
		    const ZZ& qdisc);
	void assign(const ZZ& qa, const ZZ& qb, const ZZ& qc, 
		    const ZZ& qd, const ZZ& qe);
  // latter calls set_roots_and_type()
        void operator=(const quartic& q);
        ZZ geta() const {return a;}
        ZZ getb() const {return b;}
        ZZ getcc() const {return c;} //NB "getc" is a standard macro
        ZZ getd() const {return d;}
        ZZ gete() const {return e;}
	void getcoeffs(ZZ& xa, ZZ& xb, ZZ& xc, ZZ& xd, ZZ& xe) const {xa=a; xb=b; xc=c; xd=d; xe=e;}
        int gettype() const {return type;}
        ZZ getI() const {return ii;}
        ZZ getJ() const {return jj;}
        ZZ getH() const {return H_invariant(a,b,c);}
        ZZ getdisc() const {return disc;}
        vector<bigcomplex> getroots(void) const {return roots;}
        void doubleup()
          {
	    b*=2; c*=4; d*=8; e*=16; ii*=16; jj*=64; disc*=4096;
         }
        vector<bigrational> rational_roots() const; // returns rational roots
        int trivial() const;     // Checks for a rational root
	long nrootsmod(long p) const;
        friend ostream& operator<<(ostream& s, const quartic& q);
	friend int new_equiv( quartic& q1, quartic& q2, int info);
	friend void qc(quartic& g,
		       const ZZ& x0,  const ZZ& y0,  const ZZ& z0,
		       Curvedata * E, Curvedata* IJ_curve, 
		       const ZZ& tr_u, const ZZ& tr_r, 
		       const ZZ& tr_s, const ZZ& tr_t,  
		       Point& P, int verbose);
	void dump(ostream& s) const
	  {
	    s<<"Coeffs: ("<<a<<","<<b<<","<<c<<","<<d<<","<<e<<")\n"
	     <<"Roots("<<roots<<"): \n"<<roots[0]<<"\n"<<roots[1]<<"\n"
	     <<roots[2]<<"\n"<<roots[3]<<"\n"
	     <<"Type = "<<type<<", I="<<ii<<", J="<<jj<<endl;
	  }
// Implementation
private:
       ZZ a,b,c,d,e; // coefficients
       vector<bigcomplex> roots; // 4 roots, created in all constructors
       int type;       // 1, 2 or 3
       ZZ ii,jj,disc;
// The following are used by new_equiv:  (NB p = -H)
       ZZ p, r, psq, asq;
       void make_zpol();
       int have_zpol;
       unsigned long equiv_code;
      public:
       unsigned long set_equiv_code(const vector<long>& plist);
};

inline ostream& operator<<(ostream& s, const quartic& q)
{
 s<<"("<<q.a<<","<<q.b<<","<<q.c<<","<<q.d<<","<<q.e<<")"<<flush;
 return s;
}

quartic make_quartic(const ZZ& a, const ZZ& b, const ZZ& c, const ZZ& d, const ZZ& e);

#endif
