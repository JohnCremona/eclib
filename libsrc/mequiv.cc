// mequiv.cc: implementation of quartic equivalence functions
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2005 John Cremona
// 
// This file is part of the mwrank package.
// 
// mwrank is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// mwrank is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with mwrank; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
// 
//////////////////////////////////////////////////////////////////////////
 
#include "mequiv.h"

//#ifdef NEW_EQUIV

int new_equiv(quartic* q1, quartic* q2, int info)
{
  if(info)
    {
      cout<<"Checking equivalence of " << *q1 << " and " << *q2 << "\n";
    }
  bigint& ii=q1->ii;
  bigint& jj=q1->jj;
  if (!(ii==q2->ii && jj==q2->jj && q1->disc==q2->disc && q1->type== q2->type))
    { 
      if (info) 
	{
	  cout << "equiv failed on first test!\n";
	  cout << "First  has I="<<q1->ii<<", J="<<q1->jj<<",";
	  cout << " disc="<<q1->disc<<", type="<<q1->type<<endl;
	  cout << "Second has I="<<q2->ii<<", J="<<q2->jj<<",";
	  cout << " disc="<<q2->disc<<", type="<<q2->type<<endl;
	}
      return 0;
    }

  if(q1->equiv_code!=q2->equiv_code)
    {
      if(info) cout << "--equiv_codes not equal\n";
      return 0;
    }

  q1->make_zpol();
  q2->make_zpol();
  bigint& p1=q1->p;      bigint& p2=q2->p;
  bigint& r1=q1->r;      bigint& r2=q2->r;
  bigint& p1sq=q1->psq;  bigint& p2sq=q2->psq;
  bigint& a1=q1->a;      bigint& a2=q2->a;
  bigint& a1sq=q1->asq;  bigint& a2sq=q2->asq;
  const bigint& a1a2=a1*a2;
  const bigint& p1p2=p1*p2;
  const bigint& p = (32*a1a2*ii + p1p2)/3;
  const bigint& s = (-256*jj*a1a2*(a1*p2+a2*p1) 
	       + 64*ii*(p1sq*a2sq+p2sq*a1sq+p1p2*a1a2)
	       - p1sq*p2sq) / 27;
  const bigint& r = r1*r2;
  if(info) cout<<"u-poly = [1,0, " << -2*p<< ", "<<-8*r<<", "<<s<<"]\n";

  // Compute the roots of "u-poly",  see if any are integral, and check:
  // The roots are sqrt(z1*w1)+sqrt(z2*w2)+sqrt(z3*w3) with some choice of signs,
  // where the wi are the z-values of the second quartic.  We compute the zi
  // from the roots of the first, then to ensure that conjugates are matched
  // we compute the wi from the zi (via, invisibly, the phi_i).

  bigcomplex* roots1 = q1->getroots();
  bigfloat xa1=I2bigfloat(a1), xa2=3*I2bigfloat(a2);
  bigcomplex rz1=xa1*(roots1[0]+roots1[1]-roots1[2]-roots1[3]);
  bigcomplex rz2=xa1*(roots1[0]-roots1[1]+roots1[2]-roots1[3]);
  bigcomplex rz3=xa1*(roots1[0]-roots1[1]-roots1[2]+roots1[3]);
  bigcomplex t = I2bigfloat(a1*p2-a2*p1), tt=1/I2bigfloat(3*a1);
  bigcomplex rzw1 = rz1*sqrt((xa2*rz1*rz1+t)*tt);
  bigcomplex rzw2 = rz2*sqrt((xa2*rz2*rz2+t)*tt);
  bigcomplex rzw3 = rz3*sqrt((xa2*rz3*rz3+t)*tt);

  bigcomplex u;
  
  for(long k=0; k<4; k++)
    {
      switch(k) {
      case 0:  u=rzw1+rzw2+rzw3; break;
      case 1:  u=rzw1+rzw2-rzw3; break;
      case 2:  u=rzw1-rzw2+rzw3; break;
      case 3:  u=rzw1-rzw2-rzw3; break;
      }
      if(is_approx_zero(imag(u)))
	{
	  bigint uu = Iround(real(u));
	  bigint uu2=uu*uu;
	  bigint fuu1 = (uu2-2*p)*uu2+s;
	  bigint fuu2 = 8*r*uu;
	  if(fuu1==fuu2) 
	    {
	      if(info) cout<<"Root u = "<<uu<<endl;
	      return 1;
	    }
	  if(fuu1==-fuu2) 
	    {
	      if(info) cout<<"Root u = "<<-uu<<endl;
	      return 1;
	    }
	}
    }

  if(info) cout<<"No integral roots"<<endl;
  return 0;
}

//#else // not using new_equiv so no need to compile this stuff

int testd(const bigint& a, const bigint& b, const bigint& c, 
	  const bigint& d, const bigint& e, const bigint& as, 
	  const bigint& bs, const bigint& cs, const bigint& ds, 
	  const bigint& es, const bigint& dd, const bigint& al, 
	  const bigint& be, const bigint& ga, const bigint& de, 
	  int info);

bigcomplex crossratio(const bigcomplex& x1,const bigcomplex& x2,const bigcomplex& x3,const bigcomplex& x4);

int rootsequiv(const quartic* q1, const quartic* q2, int i, const vector<bigint>& dlist, int info);

int allperms[24][4] =
                 {{0,1,2,3},{1,0,2,3},{0,1,3,2},{1,0,3,2},  // Up to here for Type III
		  {2,3,0,1},{2,3,1,0},{3,2,0,1},{3,2,1,0},  // Up to here for Type I
		  {0,2,1,3},{0,2,3,1},{0,3,1,2},{0,3,2,1},
                  {1,2,0,3},{1,2,3,0},{1,3,0,2},{1,3,2,0},
                  {2,0,1,3},{2,0,3,1},{2,1,0,3},{2,1,3,0},
                  {3,0,1,2},{3,0,2,1},{3,1,0,2},{3,1,2,0},}; // All for Type II


int testd(const bigint& a, const bigint& b, const bigint& c, 
	  const bigint& d, const bigint& e, const bigint& as, 
	  const bigint& bs, const bigint& cs, const bigint& ds, 
	  const bigint& es, const bigint& dd, const bigint& al, 
	  const bigint& be, const bigint& ga, const bigint& de, 
	  int info)
{
 bigint d2 = dd*dd;
 bigint al2=al*al, al3=al2*al, al4=al3*al;
 bigint ga2=ga*ga, ga3=ga2*ga, ga4=ga3*ga;
 bigint temp = ga4*es + al*ga3*ds + al2*ga2*cs + al3*ga*bs + al4*as - d2*a;
 if (!is_zero(temp)) return 0;
 bigint de2=de*de, de3=de2*de, de4=de3*de;
 bigint be2=be*be, be3=be2*be, be4=be3*be;
 temp = de4*es + be*de3*ds + be2*de2*cs + be3*de*bs + be4*as - d2*e;
 if (!is_zero(temp)) return 0;
 temp = 4*ga3*de*es + (3*al*ga2*de+be*ga3)*ds
                    + 2*(al2*ga*de+al*be*ga2) * cs
                    + (3*al2*be*ga+al3*de)*bs + 4*al3*be*as - d2*b;
 if (!is_zero(temp)) return 0;
 temp = 4*ga*de3*es + (3*be*ga*de2+al*de3)*ds
                   + 2*(be2*ga*de+al*be*de2)*cs
                   + (be3*ga+ 3*al*be2*de)*bs + 4*al*be3*as - d2*d;
 if (!is_zero(temp)) return 0;
 temp = 6*ga2*de2*es + 3*(be*ga2*de+al*ga*de2) * ds
                + (be2*ga2+ 4*al*be*ga*de+al2*de2) * cs
                + 3*(al*be2*ga+al2*be*de) * bs + 6*al2*be2*as - d2*c;
 if (!is_zero(temp)) return 0;
 return 1;
}  /* end of testd() */


bigcomplex crossratio(const bigcomplex& x1,const bigcomplex& x2,const bigcomplex& x3,const bigcomplex& x4)
{
 return ((x1-x3)*(x2-x4))/((x1-x4)*(x2-x3));  
}

int rootsequiv(const quartic* q1, const quartic* q2, int i, const vector<bigint>& dlist, int info)
{  int ans=0;
   bigcomplex *x = q1->getroots();
   bigcomplex *y = q2->getroots();
   bigcomplex x1=x[0], x2=x[1], x3=x[2], x4=x[3];
   bigcomplex y1=y[allperms[i][0]],y2=y[allperms[i][1]],
           y3=y[allperms[i][2]],y4=y[allperms[i][3]];
//cout<<"X-roots ("<<x<<"): "<<x1<<x2<<x3<<x4<<endl;
//cout<<"Y-roots ("<<y<<"): "<<y1<<y2<<y3<<y4<<endl;
   bigcomplex xtheta = crossratio(x1,x2,x3,x4);   
   bigcomplex ytheta = crossratio(y1,y2,y3,y4);
   if (abs(xtheta-ytheta)>0.1)
     {if (info)
        {cout << i+1 << ": Cross-ratios unequal: "<<xtheta<<" and "<<ytheta<<".\n";
       }
      return 0;
    }
   if(info) cout << i+1 << ": Cross-ratios equal: "<<xtheta<<" and "<<ytheta<<".\n";
   bigcomplex xy1=x1*y1, xy2=x2*y2,
   xy3=x3*y3, // xy4=x4*y4,
   xy23=(x2*y3)-(x3*y2),
   xy13=(y1*x3)-(x1*y3),
   xy12=(x1*y2)-(y1*x2);
   bigcomplex calpha = (xy1*(y2-y3)) + (xy2*(y3-y1)) + (xy3*(y1-y2)),
           cbeta  = (xy1*xy23) + (xy2*xy13) + (xy3*xy12),
           cgamma = xy13 + xy12 + xy23,
           cdelta = (xy1*(x2-x3)) + (xy2*(x3-x1)) + (xy3*(x1-x2));
   bigcomplex scale = calpha;
   if (abs(scale)<abs(cbeta)) scale=cbeta;
   if (abs(scale)<abs(cgamma)) scale=cgamma;
   if (abs(scale)<abs(cdelta)) scale=cdelta;
   if(is_small(scale)) 
     {
       cout << "Warning from rootsequiv(): scale = " << scale << endl;
       cout << "alpha, beta, gamma, delta = " << calpha << cbeta << cgamma << cdelta << endl;
     }
   calpha /= scale;
   cbeta  /= scale;
   cgamma /= scale;
   cdelta /= scale;
   if (!(is_real(calpha) && is_real(cbeta) 
         && is_real(cgamma) 
         && is_real(cdelta)))
     {if (info) 
	{cout << "Transformation not real.\n";
	 cout << "alpha, beta, gamma, delta = " << calpha << cbeta << cgamma << cdelta << endl;
       }
      return 0;
    }
   bigfloat alpha = real(calpha),
          beta  = real(cbeta),
          gamma = real(cgamma),
          delta = real(cdelta);
   bigfloat det = alpha*delta-beta*gamma;

   if (info) 
     {
       cout << "Real transformation has alpha, beta, gamam, delta = ";
       cout <<alpha<<" "<<beta<<" "<<gamma<<" "<<delta<<endl;
       cout << "Testing divisors of "<<q1->getdisc()<<":\n";
//       cout << dlist << endl;
     }
   bigint d;
   vector<bigint>::const_iterator dvar;
   for (dvar=dlist.begin(); dvar!=dlist.end() && (!ans); dvar++)
     {
       d = *dvar;
       bigfloat rscale = sqrt(abs(I2bigfloat(d)/det));
       bigfloat rscaler = floor(rscale+0.5);
       if(abs(rscale-rscaler)<0.001)
	 {
	   bigint al = Iround(rscale*alpha),
	   be = Iround(rscale*beta ),
	   ga = Iround(rscale*gamma),
	   de = Iround(rscale*delta);
	   bigint ddet = abs(al*de-be*ga);
	   if(d==ddet)
	     {
	       if (info)
		 {cout << "d = " << d << endl;
		  cout<<"rscale = "<<rscale<<endl;
		  cout << "al,be,ga,de = "<<al<<" "<<be<<" "<<ga<<" "<<de<<endl;
		}
	       ans=testd(q1->geta(),q1->getb(),q1->getcc(),q1->getd(),q1->gete(),
			 q2->geta(),q2->getb(),q2->getcc(),q2->getd(),q2->gete(),
			 d,al,be,ga,de,info);
	     }
	 }
     }
   return ans ;
 }    // of rootsequiv()
 
 
int equiv(const quartic* q1, const quartic* q2, const vector<bigint>& dlist, int info)
{
   bigint iiq1 = q1->getI(), jjq1 = q1->getJ(), discq1 = q1->getdisc();
   bigint iiq2 = q2->getI(), jjq2 = q2->getJ(), discq2 = q2->getdisc();
   int typeq1 = q1->gettype(), typeq2 = q2->gettype();
   if(info)
     {
       cout<<"Checking equivalence of \n"; q1->dump(cout);
       cout<<"and\n"; q2->dump(cout);
     }
   if (iiq1==iiq2 && jjq1==jjq2 && discq1==discq2 && typeq1== typeq2)
     {
       int nperms = (typeq1==1)? 8 : (typeq1==2)? 24 : 4;
       if (info) cout << "Params agree; calling rootsequiv "<<nperms<<" times.\n";
       int i,ans = 0;
       for (i=0; i<nperms && (!ans); i++)
	 {
	   ans=rootsequiv(q1,q2,i,dlist,info);
	 }
       if(info) {if(!ans) cout << "Not "; cout<<"equiv\n";}
       return ans;
     }
   else 
     { 
       if (info) {cout << "equiv failed on first test!\n";
                  cout << "First  has I="<<iiq1<<", J="<<jjq1<<",";
                  cout << " disc="<<discq1<<", type="<<typeq1<<endl;
                  cout << "Second has I="<<iiq2<<", J="<<jjq2<<",";
                  cout << " disc="<<discq2<<", type="<<typeq2<<endl;
                }
       return 0;
     }
 }  // of equiv()


//#endif
