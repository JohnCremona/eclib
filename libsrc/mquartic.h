// mquartic.h:   Declaration of class quartic and related functions
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2005 John Cremona
// 
// This file is part of the mwrank package.
// 
// mwrank is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// mwrank is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with mwrank; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
// 
//////////////////////////////////////////////////////////////////////////
 
#ifndef _MQUARTIC_H
#define _MQUARTIC_H

#define NEW_EQUIV
#include "compproc.h"
#include "marith.h"
#include "points.h"
#include "realroots.h"

inline bigint II(const bigint& a, const bigint& b, const bigint& c, 
	  const bigint& d, const bigint& e)
{
  return 12*a*e - 3*b*d + c*c;
}

inline bigint JJ(const bigint& a, const bigint& b, const bigint& c, 
	  const bigint& d, const bigint& e)
{
  return (72*a*e + 9*b*d - 2*sqr(c)) * c - 27*(a*sqr(d) + sqr(b)*e);
}

inline bigint H_invariant(const bigint& a, const bigint& b, const bigint& c)
{
  return 8*a*c - 3*sqr(b);
}

inline bigint R_invariant(const bigint& a, const bigint& b, const bigint& c, 
		 const bigint& d)
{
  return pow(b,3) + 8*d*sqr(a) - 4*a*b*c;
}


class quartic {
public:
     // constructors (NOT inline as they all use "new"
        quartic();
        quartic(const bigint& qa, const bigint& qb, const bigint& qc, 
		const bigint& qd, const bigint& qe, 
                bigcomplex* qr,	int qt,
		const bigint& qi,const bigint& qj,const bigint& qdisc);
        quartic(const bigint& qa, const bigint& qb, const bigint& qc, 
		const bigint& qd, const bigint& qe);
  // latter calls set_roots_and_type()
	~quartic();
        quartic(const quartic& q);
  // member functions & operators
        void set_roots_and_type();
	void assign(const bigint& qa, const bigint& qb, const bigint& qc, 
		    const bigint& qd, const bigint& qe, 
		    bigcomplex *qr,    int qt,
		    const bigint& qi,const bigint& qj,
		    const bigint& qdisc);
	void assign(const bigint& qa, const bigint& qb, const bigint& qc, 
		    const bigint& qd, const bigint& qe);
  // latter calls set_roots_and_type()
        void operator=(const quartic& q);
        bigint geta() const {return a;}
        bigint getb() const {return b;}
        bigint getcc() const {return c;} //NB "getc" is a standard macro
        bigint getd() const {return d;}
        bigint gete() const {return e;}
	void getcoeffs(bigint& xa, bigint& xb, bigint& xc, bigint& xd, bigint& xe) const {xa=a; xb=b; xc=c; xd=d; xe=e;}
        int gettype() const {return type;}
        bigint getI() const {return ii;}
        bigint getJ() const {return jj;}
        bigint getH() const {return H_invariant(a,b,c);}
        bigint getdisc() const {return disc;}
        bigcomplex* getroots(void) const {return roots;}
        void doubleup()
          {
	    b*=2; c*=4; d*=8; e*=16; ii*=16; jj*=64; disc*=4096;
         }
        int trivial() const;     // Checks for a rational root
	long nrootsmod(long p) const;
        friend ostream& operator<<(ostream& s, const quartic& q);
	friend int new_equiv(quartic* q1, quartic* q2, int info);
	friend void qc(quartic& g,
		       const bigint& x0,  const bigint& y0,  const bigint& z0,
		       Curvedata * E, Curvedata* IJ_curve, 
		       const bigint& tr_u, const bigint& tr_r, 
		       const bigint& tr_s, const bigint& tr_t,  
		       Point& P, int verbose);
	void dump(ostream& s) const
	  {
	    s<<"Coeffs: ("<<a<<","<<b<<","<<c<<","<<d<<","<<e<<")\n"
	     <<"Roots("<<roots<<"): \n"<<roots[0]<<"\n"<<roots[1]<<"\n"
	     <<roots[2]<<"\n"<<roots[3]<<"\n"
	     <<"Type = "<<type<<", I="<<ii<<", J="<<jj<<endl;
	  }
// Implementation
private:
       bigint a,b,c,d,e; // coefficients
       bigcomplex* roots; // roots, array 0f 4 created in all constructors
       int type;       // 1, 2 or 3
       bigint ii,jj,disc;
// The following are used by new_equiv:  (NB p = -H)
       bigint p, r, psq, asq;
       void make_zpol();
       int have_zpol;
       unsigned long equiv_code;
      public:
       unsigned long set_equiv_code(const vector<long>& plist);
};

inline ostream& operator<<(ostream& s, const quartic& q)
{
 s<<"("<<q.a<<","<<q.b<<","<<q.c<<","<<q.d<<","<<q.e<<")"<<flush;
 return s;
}

quartic make_quartic(const bigint& a, const bigint& b, const bigint& c, const bigint& d, const bigint& e);

#endif
