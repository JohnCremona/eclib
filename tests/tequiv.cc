// tequiv.cc: test program for quartic equivalence
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
//
#include <eclib/mquartic.h>
#include <eclib/mequiv.h>
#define NEQPLIST 0        // Number of primes for equiv-test sieving

vector<long> eqplist;

int getquartic(quartic& g, int verbose)
{
  ZZ a, b, c, d, e;
  
  if(verbose)  cout << "Enter quartic coefficients a,b,c,d,e ?" << endl;
  char ch; cin>>ch;
  if(ch=='(') cin>>a>>ch>>b>>ch>>c>>ch>>d>>ch>>e>>ch;
     else 
     {
       cin.putback(ch);
       cin >> a >> b >> c >> d >> e;
     }
     
  if (is_zero(a)&&is_zero(b)&&is_zero(c)&&is_zero(d)&&is_zero(e))
     return 0;
    
  g=quartic(a,b,c,d,e);  // will set its own invariants, roots and type
  g.set_equiv_code(eqplist);
  return 1;
}

int main()
{
  initprimes("PRIMES",0);
  cout.precision(50);
  cin.flags( cin.flags() | ios::dec );
  
  int verb; cout << "Verbose? "; cin >> verb;
  while (1)
    {
  long nq;  cout << "How many quartics to check (0 to quit)? ";  cin >> nq;
  cout<<endl<<endl;
  if (nq<1) {return 0;}
  vector<quartic> glist(nq);
  vector<ZZ> dlist;
  int i,j;

  for(i=0; i<nq; i++)
    {
      getquartic(glist[i], verb);
      cout<<(i+1)<<": "<<glist[i];
      if(verb) cout<<endl;
      int eq=0;
      for(j=0; (j<i)&&(!eq); j++)
	{
	  eq = new_equiv(glist[i],glist[j],verb);
	  if(eq) cout<<" equivalent to #"<<(j+1)<<endl;
	}
      if(!eq) cout<<" new"<<endl;
    }
    }
  return(0);
}
