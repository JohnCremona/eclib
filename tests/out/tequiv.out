Verbose? How many quartics to check? 

Enter quartic coefficients a,b,c,d,e ?
1: (1,0,-80,324,-244)
 new
Enter quartic coefficients a,b,c,d,e ?
2: (1,0,-68,236,-96)
Checking equivalence of (1,0,-68,236,-96) and (1,0,-80,324,-244)
u-poly = [1,0, -306176, -39149568, 8727953408]
No integral roots
 new
Enter quartic coefficients a,b,c,d,e ?
3: (1,1,-35,11,190)
Checking equivalence of (1,1,-35,11,190) and (1,0,-80,324,-244)
u-poly = [1,0, -194816, -4748544, 7849278464]
No integral roots
Checking equivalence of (1,1,-35,11,190) and (1,0,-68,236,-96)
u-poly = [1,0, -176704, -3458816, 6658838528]
Root u = -272
 equivalent to #2
Enter quartic coefficients a,b,c,d,e ?
4: (1,0,-26,12,233)
Checking equivalence of (1,0,-26,12,233) and (1,0,-80,324,-244)
equiv failed on first test!
First  has I=3472, J=-404912, disc=3463112448, type=1
Second has I=3472, J=-404912, disc=3463112448, type=2
Checking equivalence of (1,0,-26,12,233) and (1,0,-68,236,-96)
equiv failed on first test!
First  has I=3472, J=-404912, disc=3463112448, type=1
Second has I=3472, J=-404912, disc=3463112448, type=2
Checking equivalence of (1,0,-26,12,233) and (1,1,-35,11,190)
equiv failed on first test!
First  has I=3472, J=-404912, disc=3463112448, type=1
Second has I=3472, J=-404912, disc=3463112448, type=2
 new
